vtex/faststore
# Invert direction of dependency between @faststore/core and @faststore/cli

## What's the purpose of this pull request?

This PR aims to make `@faststore/core` a dependency of the `@faststore/cli`. It just makes sense. Someone can fork `@faststore/core` and never need to use the CLI, but there's hardly any usefulness to the CLI without `@faststore/core`.

## How it works?

The only thing linking Core to the CLI was the `generate-graphql` command on the CLI. It just called back 4 scripts on core's `package.json`: `generate:schema`, `generate:codegen`, `format:generated` and `generate:copy-back`. The copy-back command was only needed to improve the DX of customers customizing their GraphQL schema [[ref]](https://github.com/vtex/faststore/pull/2176#discussion_r1438436193) and it was the only part of the script that needed information available at the CLI level only.

Since it was only a DX improvement, all that was needed was, during the `faststore dev` execution I had to generate the new schema, copy the files to the correct `node_modules` dir, and then start the NextJS dev server. In order to do this I split the `dev` script into `predev`, `dev`, and `dev-only`*. From the FastStore CLI I could now call `yarn predev`, copy the files to the correct place, and then run `yarn dev-only` to start the NextJS dev server.

Someone developing on the faststore monorepo (us VTEX developers, or someone contributing to our project) will continue to run core via `yarn dev`. [npm](https://docs.npmjs.com/cli/v10/using-npm/scripts#pre--post-scripts), [yarn](https://classic.yarnpkg.com/en/docs/cli/run#toc-yarn-run-script), and [pnpm](https://pnpm.io/cli/run#npmrc-settings) support the lifecycle scripts out of the box. I did the same thing to `build` for equivalence.

One thing that can be up to discussion is the existence of `runCommandSync` util. It was only used on `generate-graphql` and can be easily removed (in this PR or in a future one).

## How to test it?

Add this PR `@faststore/cli` from codesandbox as a dependency of the starter store, remove `@faststore/core` as an explicit dependency, and `@faststore/cli` as a devDependency. Run your store and see that everything still works fine. Or go to the [starter preview](https://sfj-20ca89a--starter.preview.vtex.app/) [[PR]](https://github.com/vtex-sites/starter.store/pull/489) here and see the store running in "production".
